hawc/apps/animal/exports.py

```diff
@@ -1,744 +1,1520 @@
+                    "NOEL",
+                    "LOEL",
+                    "FEL",
+                    "variance",
+                ),
+            ),
+            DoseGroupExport(
+                "dose_group",
+                "animal_group__dosing_regime__doses",
+                include=("id", "dose_units_id", "dose_units_name", "dose_group_id", "dose"),
+            ),
+        ]
 
 
 class EndpointGroupFlatDataPivot(FlatFileExporter):
-    """
-    Return a subset of frequently-used data for generation of data-pivot
-    visualizations.
-    """
-
-    @classmethod
-    def _get_doses_list(cls, ser, preferred_units):
-        # compact the dose-list to only one set of dose-units; using the
-        # preferred units if available, else randomly get first available
-        units_id = None
-
+    def get_preferred_units(self, df: pd.DataFrame) -> int | None:
+        preferred_units = self.kwargs.get("preferred_units", None)
+        available_units = df["dose_group-dose_units_id"].dropna().unique()
+        if available_units.size == 0:
+            return None
         if preferred_units:
-            available_units = set(
-                [d["dose_units"]["id"] for d in ser["animal_group"]["dosing_regime"]["doses"]]
-            )
             for units in preferred_units:
                 if units in available_units:
-                    units_id = units
-                    break
+                    return units
+        return available_units[0]
+
+    def handle_ci(self, df: pd.DataFrame) -> pd.DataFrame:
+        def _func(row: pd.Series) -> pd.Series:
+            # logic used from EndpointGroup.getConfidenceIntervals()
+            data_type = row["endpoint-data_type"]
+            lower_ci = row["endpoint_group-lower_ci"]
+            upper_ci = row["endpoint_group-upper_ci"]
+            n = row["endpoint_group-n"]
+
+            response = row["endpoint_group-response"]
+            stdev = row["endpoint_group-stdev"]
+            incidence = row["endpoint_group-incidence"]
+            if lower_ci is not None or upper_ci is not None or n is None or n <= 0:
+                pass
+            elif (
+                data_type == constants.DataType.CONTINUOUS
+                and response is not None
+                and stdev is not None
+            ):
+                (
+                    row["endpoint_group-lower_ci"],
+                    row["endpoint_group-upper_ci"],
+                ) = cont_ci(stdev, n, response)
+            elif (
+                data_type in [constants.DataType.DICHOTOMOUS, constants.DataType.DICHOTOMOUS_CANCER]
+                and incidence is not None
+            ):
+                (
+                    row["endpoint_group-lower_ci"],
+                    row["endpoint_group-upper_ci"],
+                ) = dich_ci(incidence, n)
+            return row
+
+        return df.apply(_func, axis="columns")
+
+    def handle_stdev(self, df: pd.DataFrame) -> pd.DataFrame:
```

Comment:
I noticed that some code was almost identical in a few places; we could refactor in the future, but it's fine to keep as is for now, since we're doing our evaluation to make sure it returns identical results.

Reply:
At the top of the file I tried to make reusable functions for some of the calculations, but there are some pandas specific code that I didn't refactor, so I agree there's some work to do there. I believe this is better for a future PR as well, since there may be refactoring we can do between export files too which would fall outside the scope of this PR.